File: examples/cpp/rx_to_file.cpp
//
// Created by Jon Kler on 6/10/25.
//

#include <mantis/mantis.h>
#include <fstream>
#include <boost/program_options.hpp>

namespace po = boost::program_options;

static constexpr double RX_BUFF_SIZE = 1e5;
static constexpr double threshold = 1e-6;
static constexpr size_t INDICATOR_SLEEP_TIME = 3;

static std::string program_description = "Receive samples at specified RX params and write to file.";

#define PASSES_THRESHOLD(a, b) (std::abs(a - b) < threshold)

int main(int argc, char **argv) {
    po::options_description desc(program_description + "\nOptions");
    desc.add_options()
            ("help,h", "produce help message")
            ("filename", po::value<std::string>()->required(),
             "REQUIRED. Full path of the file to write to. Opens in overwrite mode")
            ("args", po::value<std::string>(), "device args str")
            ("freq", po::value<double>()->default_value(100e6, "100e6"), "Rx freq [Hz], will default to 100MHz")
            ("lo", po::value<double>(), "lo offset freq [Hz] By default will allow SDR lib to choose most fitting lo")
            ("rate", po::value<double>()->default_value(10e6, "10e6"), "Rx rate [Sps], will default to 10 MSps")
            ("gain", po::value<double>()->default_value(0.0), "Rx gain")
            ("master_clock_rate,mcr", po::value<double>(),
             "master clock rate; can also be passed in args as master_clock_rate. This argument takes precedence")
            ("channel", po::value<size_t>()->default_value(0), "channel num");

    po::variables_map vm;
    po::store(po::parse_command_line(argc, argv, desc), vm);

    if (vm.count("help")) {
        std::cout << desc << "\n";
        return EXIT_SUCCESS;
    }

    po::notify(vm);


    mantis::utils::pversion();
    
    /// first, try to open file
    std::ofstream file(vm["filename"].as<std::string>(), std::ios::out | std::ios::binary);
    if (!file.is_open()) {
        mantis::utils::perror("Failed to open file: " + vm["filename"].as<std::string>());
        return EXIT_FAILURE;
    }

    auto &d_manager = mantis::device_manager::get_instance();

    /// acquire sdr
    mantis::params::msdr_params params{};
    if (vm.count("args")) {
        if (!mantis::errors::succeeded(mantis::params::msdr_params::from_str(vm["args"].as<std::string>(), params))) {
            throw mantis::runtime_error("Could not parse arg string: " + vm["args"].as<std::string>());
        }
    }

    if (vm.count("master_clock_rate")) {
        params.master_clock_rate = vm["master_clock_rate"].as<double>();
    }

    if (!mantis::errors::succeeded(d_manager.init(1, params))) {
        throw mantis::runtime_error("Could not init relevant SDR");
    }

    std::optional<double> lo = (vm.count("lo") ? std::optional<double>{vm["lo"].as<double>()} : std::nullopt);

    double freq = vm["freq"].as<double>();
    double rate = vm["rate"].as<double>();
    double gain = vm["gain"].as<double>();
    size_t channel = vm["channel"].as<size_t>();

    /// try to acquire channel
    auto [err, rx_channel] = d_manager.get_rx_channel(params, channel);
    if (!mantis::errors::succeeded(err)) {
        mantis::utils::perror("Failed to Acquire Channel: " + mantis::errors::mantis_errno(err));
    }

    /// configure channel
    double ac_freq = rx_channel->set_freq(freq, lo);
    if (!PASSES_THRESHOLD(ac_freq, freq)) {
        mantis::utils::pwarn(
            "Couldn't set freq to: " + std::to_string(freq) + ". Actual freq: " + std::to_string(ac_freq) + " Hz");
    }

    double ac_gain = rx_channel->set_gain(gain);
    if (!PASSES_THRESHOLD(ac_gain, gain)) {
        mantis::utils::pwarn(
            "Couldn't set gain to: " + std::to_string(gain) + ". Actual gain: " + std::to_string(ac_gain) + " dB");
    }

    double ac_rate = rx_channel->set_rate(rate);
    if (!PASSES_THRESHOLD(ac_rate, rate)) {
        mantis::utils::pwarn(
            "Couldn't set rate to: " + std::to_string(rate) + ". Actual rate: " + std::to_string(ac_rate) + " Hz");
    }

    char *rx_buff = (char *) malloc(RX_BUFF_SIZE * sizeof(char));
    if (!rx_buff) {
        mantis::utils::perror("Failed to malloc Rx buffer");
        return EXIT_FAILURE;
    }

    int sample_size;
    if (mantis::config::CPU_SAMPLE_FORMAT == "fc32") {
        sample_size = 8;
    } else if (mantis::config::CPU_SAMPLE_FORMAT ==
               "fc64") {
        // should also not be mantis api, consider compiling with diff library target for these
        sample_size = 16;
    } else {
        mantis::utils::perror("Unknown sample formant");
        return EXIT_FAILURE;
    }
    (void) sample_size;

    mantis::go::mrx_metadata rx_md{.start_of_burst = true};

    mantis::utils::pinfo("++++++Begin RX+++++");
    mantis::utils::pinfo("Quit to stop");

    size_t sample_num = RX_BUFF_SIZE / sample_size;

    // dispatch thread to indicate rx
    bool running = true;
    auto indicator = std::thread([&running]() {
        while (running) {
            std::this_thread::sleep_for(std::chrono::seconds(INDICATOR_SLEEP_TIME));
            std::cout << "+" << std::flush;
        }
    });

    while (true) {
        if (!rx_channel->is_valid()) [[unlikely]] {
            mantis::utils::perror("SDR healthcheck Failed, channel invalid. Shutting Down...");
            running = false;
            indicator.join();
            return EXIT_FAILURE;
        }

        rx_channel->receive(rx_buff, sample_size, sample_num, rx_md);
        rx_md.has_time_spec = false; // reset timespec so we dont schedule the next recv to be in the past 
        file.write(rx_buff, sample_num);
    }
}

File: examples/python/rx_to_file.py
import mantis
from mantis import errors as me
from mantis import utils as mu
import argparse
import numpy as np

RX_BUFF_SIZE = 1e5
THRESHOLD = 1e-6


def passes_threshold(val_1, val_2):
    return abs(val_1 - val_2) < THRESHOLD


def main():
    parser = argparse.ArgumentParser(description="Rx to file")
    parser.add_argument(
        "-a",
        "--args",
        type=str,
        default="",
        help="device args str. Leave empty to find all",
    )
    parser.add_argument(
        "-f",
        "--filename",
        type=str,
        default=None,
        help="absolute path of the file to transmit",
    )

    parser.add_argument("--freq", type=float, default=100e6, help="frequency in [Hz]")
    parser.add_argument("--gain", type=float, default=0, help="gain in [dB]")
    parser.add_argument("--rate", type=float, default=10e6, help="Rx rate in [Sps]")
    parser.add_argument(
        "--lo",
        type=float,
        default=None,
        help="lo offset freq [Hz] By default will allow SDR lib to choose most fitting lo",
    )
    parser.add_argument("-c", "--channel_num", type=int, default=0, help="channel num")
    parser.add_argument(
        "-mcr",
        "--master_clock_rate",
        type=str,
        default=None,
        help="master clock rate; can also be passed in args as master_clock_rate. "
             "This argument takes precedence",
    )

    args = parser.parse_args()

    if args.filename is None:
        mu.perror("Please provide a file to receive into")
        return

    d_manager = mantis.device_manager.get_instance()

    err, params = mantis.msdr_params.from_str(args.args)
    if err != me.error_code.SUCCESS:
        mu.perror(
            f"Failed to parse msdr_params from args string: {args.args} -- {me.mantis_errno(err)}"
        )
        return

    if args.master_clock_rate is not None:
        params.master_clock_rate = args.master_clock_rate

    freq = args.freq
    gain = args.gain
    rate = args.rate
    channel_num = args.channel_num
    lo = args.lo

    # init sdr
    err = d_manager.init(1, params)
    if err != me.error_code.SUCCESS:
        mu.perror("Could not init relevant SDR: " + me.mantis_errno(err))
        return

    # acquire channel
    err, rx_channel = d_manager.get_rx_channel(params, channel_num)
    if err != me.error_code.SUCCESS:
        mu.perror("Failed to acquire RX channel: " + me.mantis_errno(err))
        return

    # configure channel
    ac_freq = rx_channel.set_freq(freq, lo)
    if not passes_threshold(ac_freq, freq):
        mu.pwarn(
            "Couldn't set freq to: "
            + str(freq)
            + ". Actual freq: "
            + str(ac_freq)
            + " Hz"
        )

    ac_gain = rx_channel.set_gain(gain)
    if not passes_threshold(ac_gain, gain):
        mu.pwarn(
            "Couldn't set gain to: "
            + str(gain)
            + ". Actual gain: "
            + str(ac_gain)
            + " dB"
        )
    ac_rate = rx_channel.set_rate(rate)
    if not passes_threshold(ac_rate, rate):
        mu.pwarn(
            "Couldn't set rate to: "
            + str(rate)
            + ". Actual rate: "
            + str(ac_rate)
            + " Sps"
        )

    sample_size = mu.get_sample_size()

    num_samples = int(RX_BUFF_SIZE // sample_size)

    buff = np.zeros(num_samples, dtype=np.complex64)
    mv_buff = memoryview(buff).cast("B")

    rx_md = mantis.mrx_metadata()
    rx_md.start_of_burst = True

    mu.pinfo("Starting Rx to File")
    with open(args.filename, "wb") as f:
        while True:
            if not rx_channel.is_valid():
                mu.perror(
                    "SDR healthcheck failed, channel is no longer valid. shutting down..."
                )
                break

            rx_channel.receive(buff, rx_md)
            rx_md.has_time_spec = False # reset timespec so we dont schedule the next recv to be in the past
            f.write(mv_buff)


if __name__ == "__main__":
    main()
